### PR TITLE
Headless/tahsin/review/create rating

### PR DESCRIPTION
**Pull request by ghost**
# Description

(Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.)

## Type of change

(Please delete options that are not relevant.)

- [x] Link to a **GH Issue** or **Jira** task here for reference.
- [ ] Bug fix
- [x] New feature

# How Has This Been Tested?

(Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration)

- [x] Tested by myself
- [ ] Unit test

# Developer Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

# Review check list:

- [x] Separation of Concerns followed (Reviewer name/NaN)
- [x] Code is in sync with existing code patterns/technologies. (Reviewer name/NaN)
- [ ] Are there any obvious optimizations that will improve performance? (Reviewer name/NaN)
- [ ] Can any logging or debugging code be removed? (Reviewer name/NaN)
- [x] Are all data inputs checked (for the correct type, length, format, and range) and encoded? (Reviewer name/NaN)
- [x] Are invalid parameter values handled? (Reviewer name/NaN)
- [x] Is error handling done the correct way? (Reviewer name/NaN)
- [x] Are all the edge cases (null, 0, negative) handled properly? (Reviewer name/NaN)

### CODE REVIEW

**Comment by ghost** [comment]
[https://pm23.atlassian.net/browse/BC-692](url)
[https://pm23.atlassian.net/browse/BC-693](url)

File: packages/headless/src/database/mongodb/order/index.ts [hunk 1]
```diff
@@ -278,6 +280,69 @@ export class OrderDatabase implements IOrderDatabase {
...
+  }
+
+  async addProductRating(productId: string, rating: number): Promise<boolean> {
+    try {
+      const product = await ProductModel.findOne({ id: productId })
+        .lean()
+        .exec();
+      if (!product) return false;
+
+      let ratingObj = product.rating || {};
```
Comment (Anjarul-Islam) [inline packages/headless/src/database/mongodb/order/index.ts:315]: rating object is initialized by an integer!? this should be fixed

File: packages/headless/src/database/mongodb/order/index.ts [hunk 2]
```diff
@@ -278,6 +280,69 @@ export class OrderDatabase implements IOrderDatabase {
...
+        noOfRatings = noOfRatings + ratingObj[key];
+        sum = sum + (parseInt(key) * ratingObj[key]);
+      }
+
+      const avgRating = Math.round(sum / noOfRatings);
+      const newProduct = { ...product, rating: ratingObj, avgRating };
+
+      const response = await ProductModel.findOneAndUpdate(
+        { id: productId },
+        newProduct,
```
Comment (Anjarul-Islam) [inline packages/headless/src/database/mongodb/order/index.ts]: don't reset the full document, rather update only those properties which should be updated

File: packages/headless/src/database/mongodb/review/review.model.ts [hunk 3]
```diff
@@ -17,35 +17,34 @@ const ReviewSchema = new Schema<Review>(
...
-          type: String,
-          default: () => randomUUID(),
-          unique: true
-        },
+        url: String,
+        _id: false
+      }
+    ],
+    reply:{
         commentedBy:{
```
Comment (Anjarul-Islam) [inline packages/headless/src/database/mongodb/review/review.model.ts]: keep admin id as reference

File: packages/headless/src/database/mongodb/order/index.ts (hunk 1 above)
Comment (Anjarul-Islam) [inline packages/headless/src/database/mongodb/order/index.ts:323]: ``let ratingObj = product.rating || {};`` 
this code can be rewritten as
``const ratingObj = product.rating || {[rating]: 1};``

File: packages/headless/src/modules/order/repositories/index.ts [hunk 4]
```diff
@@ -113,4 +115,16 @@ export class OrderRepository {
   ): Promise<OrderEntity[]> {
     return await this.db.getOrderList(query, skip, limit);
   }
+
+  async createReview(review: any):Promise<Review | null>{
+    return await this.db.createReview(review);
+  }
+
+  async findReview(query: Record<string,any>): Promise<Review[] | null>{
+    return await this.db.findReview(query);
```
Comment (Anjarul-Islam) [inline packages/headless/src/modules/order/repositories/index.ts:124]: use pagination here (should be required)

File: packages/headless/src/modules/order/rest/review.controller.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+@Controller('order/review')
+export class OrderReviewController {
+    constructor(private orderReviewService: OrderReviewService) {}
+
+    @Post()
+    async createReview(
+        @Body() body: CreateReviewDto,
+        @Res({ passthrough: true }) res: Response,
+    ){
+        const {code, ...response} = await this.orderReviewService.createReview(body);
```
Comment (Anjarul-Islam) [inline packages/headless/src/modules/order/rest/review.controller.ts:19]: review should be taken once. a customer will be able to create a single review for an order

File: packages/headless/src/modules/order/services/review.service.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+                error: CreateReviewErrorMessage.CANNOT_CREATE_REVIEW,
+                errors: null,
+                code: HttpStatus.INTERNAL_SERVER_ERROR
+            };
+
+        return { code: 200, data: review };
+    }
+
+    async getProductReview(productId: string): Promise<ProductReviewResponse>{
+        const reviews = await this.orderRepository.findReview({ productId });
```
Comment (Anjarul-Islam) [inline packages/headless/src/modules/order/services/review.service.ts]: use pagination here.

**Comment by ghost** [comment]
[https://pm23.atlassian.net/browse/BC-717](url)

**Pull request merged**